### PR TITLE
FEATURE: Add bulk destroy to admin users list

### PR DESCRIPTION
**Pull request by OsamaSayegh**
This PR introduces a new feature that allows staff to bulk select and delete users directly from the users list at `/admin/users/list`. The main use-case for this feature is make deleting spammers easier when a site is under a large spam attack.

Screenshot:

Internal topic: t/140321.

### CODE REVIEW

File: app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js [hunk 1]
```diff
@@ -106,4 +115,54 @@ export default class AdminUsersListShowController extends Controller.extend(
       asc,
     });
   }
+
+  @action
+  toggleBulkSelect() {
+    this.bulkSelect = !this.bulkSelect;
+    this.displayBulkActions = false;
+    this.bulkSelectedUsers = null;
```
[the remark this answers is not in the document]
Comment (OsamaSayegh) [reply app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js]: Done!

File: app/assets/javascripts/admin/addon/components/bulk-user-delete-confirmation.gjs [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+            username: data.username,
+          }
+        )
+      );
+    } else if (data.failed) {
+      this.failedUsernames.push(data.username);
+      this.#logError(
+        i18n(
+          "admin.users.bulk_actions.delete.confirmation_modal.user_delete_failed",
+          {
```
Comment (martin-brennan) [inline app/assets/javascripts/admin/addon/components/bulk-user-delete-confirmation.gjs:80]: Could just pass `data` as an argument to `i18n` here instead of specifying which keys, I don't think it would error

File: app/assets/javascripts/admin/addon/components/bulk-user-delete-confirmation.gjs [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    });
+  }
+
+  @bind
+  onDeleteProgress(data) {
+    if (data.success) {
+      this.#logSuccess(
+        i18n(
+          "admin.users.bulk_actions.delete.confirmation_modal.user_delete_succeeded",
+          {
```
Comment (martin-brennan) [inline app/assets/javascripts/admin/addon/components/bulk-user-delete-confirmation.gjs:68]: Same data argument comment here

File: app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js [hunk 4]
```diff
@@ -1,25 +1,34 @@
...
   CanCheckEmails
 ) {
-  model = null;
+  @service dialog;
+  @service modal;
+
+  @tracked bulkSelect = false;
+  @tracked displayBulkActions = false;
+  @tracked bulkSelectedUserIdsSet = new Set();
+  @tracked bulkSelectedUsersMap = {};
```
Comment (martin-brennan) [inline app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js:23]: You might be able to use a JS `Map` here instead

File: app/services/user/bulk_destroy.rb [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  end
+
+  model :users
+  policy :can_delete_users
+  step :delete
+
+  private
+
+  def fetch_users(params:)
+    ids = params.user_ids.to_a
```
Comment (martin-brennan) [inline app/services/user/bulk_destroy.rb:19]: Why do you need to call to_a here since the attribute is already an :array ?

File: app/services/user/bulk_destroy.rb [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    # order as they're selected in the UI
+    User.where(id: ids).order(DB.sql_fragment("array_position(ARRAY[?], users.id)", ids))
+  end
+
+  def can_delete_users(guardian:, users:)
+    users.all? { |u| guardian.can_delete_user?(u) }
+  end
+
+  def delete(users:, guardian:)
+    users.each.with_index do |u, index|
```
Comment (martin-brennan) [inline app/services/user/bulk_destroy.rb:31]: Might want to do `find_each` here so the users are loaded in memory in batches

File: app/services/user/bulk_destroy.rb [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+      position = index + 1
+      success =
+        UserDestroyer.new(guardian.user).destroy(
+          u,
+          delete_posts: true,
+          prepare_for_destroy: true,
+          context: I18n.t("staff_action_logs.bulk_user_delete", users: users.map(&:id).inspect),
```
Comment (martin-brennan) [inline app/services/user/bulk_destroy.rb:38]: I think it's unnecessary to add all the user IDs as context, there is a lot of repetition and also you are doing map for every single user in the array

File: app/services/user/bulk_destroy.rb (hunk 6 above)
Comment (martin-brennan) [inline app/services/user/bulk_destroy.rb:31]: Also nitpicky but can you use `user` instead of `u` here to help with readability, in small blocks I think it's fine but in bigger ones like this I find it makes things harder

File: app/services/user/bulk_destroy.rb [hunk 8]
```diff
@@ -0,0 +1,69 @@
+# frozen_string_literal: true
+
+class User::BulkDestroy
```
Comment (martin-brennan) [inline app/services/user/bulk_destroy.rb:3]: Love that you did a service :clap:

File: app/assets/javascripts/admin/addon/components/bulk-user-delete-confirmation.gjs [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+  }
+
+  get confirmDeletePhrase() {
+    return i18n(
+      "admin.users.bulk_actions.delete.confirmation_modal.confirmation_phrase",
+      { count: this.args.model.userIds.length }
+    );
+  }
+
+  #logError(line) {
```
Comment (martin-brennan) [inline app/assets/javascripts/admin/addon/components/bulk-user-delete-confirmation.gjs:43]: I really like the logging UI here, I feel like it could be its own component. We do something like this already for backups and for site updates, is there any way these could be reused or for this one to supersede them in a fancy new component (in a separate PR of course)?

File: app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js [hunk 10]
```diff
@@ -106,4 +116,42 @@ export default class AdminUsersListShowController extends Controller.extend(
...
+  @action
+  toggleBulkSelect() {
+    this.bulkSelect = !this.bulkSelect;
+    this.displayBulkActions = false;
+    this.bulkSelectedUsersMap = {};
+    this.bulkSelectedUserIdsSet = new Set();
+  }
+
+  @action
+  bulkSelectItemToggle(userId, event) {
```
Comment (martin-brennan) [inline app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js:129]: Would be good to enforce the 100 user max here as well, to save a trip to the server for the user if they click more than 100. Just pop a toast to say the max amount of users for bulk actions has been reached

File: app/assets/javascripts/admin/addon/components/bulk-user-delete-confirmation.gjs [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+            position: data.position,
+            total: data.total,
+            username: data.username,
+            error: data.error,
+          }
+        )
+      );
+    }
+
+    if (data.position === data.total) {
```
Comment (martin-brennan) [inline app/assets/javascripts/admin/addon/components/bulk-user-delete-confirmation.gjs:90]: Just wondering what happens here if the bulk action never finishes/times out? Maybe we need some `setTimeout` or something in this code to handle that. Or something to handle the `hijack` timeout from the server maybe?

**Review by martin-brennan — COMMENTED** [review]
This is looking super nice so far Osama :ok_hand:

File: app/services/user/bulk_destroy.rb (hunk 5 above)
Comment (OsamaSayegh) [reply app/services/user/bulk_destroy.rb]: This was copied from an existing service 😅 I've removed it and it seems to work

File: app/services/user/bulk_destroy.rb (hunk 6 above)
Comment (OsamaSayegh) [reply app/services/user/bulk_destroy.rb]: done!

File: app/services/user/bulk_destroy.rb (hunk 7 above)
Comment (OsamaSayegh) [reply app/services/user/bulk_destroy.rb]: oh my bad, removed

File: app/services/user/bulk_destroy.rb (hunk 6 above)
Comment (OsamaSayegh) [reply app/services/user/bulk_destroy.rb]: done!

File: app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js (hunk 10 above)
Comment (OsamaSayegh) [reply app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js:133]: done

File: app/assets/javascripts/admin/addon/components/bulk-user-delete-confirmation.gjs (hunk 9 above)
Comment (OsamaSayegh) [reply app/assets/javascripts/admin/addon/components/bulk-user-delete-confirmation.gjs:43]: Possibly. My main concern with this implementation would be performance — I don't know how well this implementation scales if used in a place where thousands of log entries could be printed. In this case, we're limited to 100 users so performance isn't a huge concern.

File: app/assets/javascripts/admin/addon/components/bulk-user-delete-confirmation.gjs (hunk 11 above)
Comment (OsamaSayegh) [reply app/assets/javascripts/admin/addon/components/bulk-user-delete-confirmation.gjs:81]: I have no idea what would happen 😅 Is there a time limit for `hijack`? I took a quick look and don't see any timeout in place.

In any case, the ajax call to `/admin/users/destroy-bulk` is wrapped in a `try`/`catch` so I think that should surface any issues that could arise with the request.

File: app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js (hunk 4 above)
Comment (OsamaSayegh) [reply app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js:26]: I was hoping that `Map` could replace both the `Set` and the POJO, but `Map` doesn't work with Ember's `get` method (see the link), so I have to stick to `Set` and POJO.

https://github.com/discourse/discourse/pull/29744/files#diff-dcc9c57dab8ff1ac002b005e803cf648b760828994fc97ad2b249c0a265236f6R184

File: app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js [hunk 12]
```diff
@@ -106,4 +120,79 @@ export default class AdminUsersListShowController extends Controller.extend(
...
+    if (event.target.checked) {
+      if (this.bulkSelectedUserIdsSet.size === MAX_BULK_SELECT_LIMIT) {
```
Comment (martin-brennan) [inline app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js:135]: I think this needs to be `>=` not `===`, otherwise you can bypass it with bulk selection

File: app/assets/javascripts/admin/addon/components/bulk-user-delete-confirmation.gjs (hunk 9 above)
Comment (martin-brennan) [reply app/assets/javascripts/admin/addon/components/bulk-user-delete-confirmation.gjs:43]: Okay, I guess we just keep an eye out for more use cases, would be nice to have some unified logger component that is performant that we can use throughout the admin UI

File: app/assets/javascripts/admin/addon/components/bulk-user-delete-confirmation.gjs [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+          this.#logNeutral(`* ${username}`);
+        }
+      }
+    }
+  }
+
+  @action
+  onPromptInput(event) {
+    this.confirmButtonDisabled =
+      event.target.value.toLowerCase() !== this.confirmDeletePhrase;
```
Comment (Drenmi) [inline app/assets/javascripts/admin/addon/components/bulk-user-delete-confirmation.gjs:104]: Slightly thrown off by the double negative, but I think it's fine, if we change this to "enabled" we need to invert that in the template, so just a trade-off to make.

File: app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js [hunk 14]
```diff
@@ -1,25 +1,38 @@
+import { tracked } from "@glimmer/tracking";
```
Comment (Drenmi) [inline app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js:1]: I think there's a lot that can be done in terms of code organization here, but since this is still a controller, we can tackle that separately as a modernization project. 🙂

File: app/assets/javascripts/admin/addon/templates/users-list-show.hbs [hunk 15]
```diff
@@ -130,14 +167,50 @@
...
+                {{else}}
+                  <DTooltip
+                    @identifier="bulk-delete-unavailable-reason"
+                    @placement="bottom-start"
+                  >
+                    <:trigger>
+                      <input
+                        type="checkbox"
+                        class="directory-table__cell-bulk-select"
+                        disabled={{true}}
```
Comment (Drenmi) [inline app/assets/javascripts/admin/addon/templates/users-list-show.hbs:197]: I guess `disabled="disabled"` would work the same?

File: app/services/user/bulk_destroy.rb [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+  policy :can_delete_users
+  step :delete
+
+  private
+
+  def fetch_users(params:)
+    ids = params.user_ids
+    # this order cluase ensures we retrieve the users in the same order as the
+    # IDs in the param. we do this to ensure the users are deleted in the same
+    # order as they're selected in the UI
```
Comment (Drenmi) [inline app/services/user/bulk_destroy.rb:22]: That's cool, but is it something we need to care about in the first place? 🙂

File: app/services/user/bulk_destroy.rb [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+    User.where(id: ids).order(DB.sql_fragment("array_position(ARRAY[?], users.id)", ids))
+  end
+
+  def can_delete_users(guardian:, users:)
```
Comment (Drenmi) [inline app/services/user/bulk_destroy.rb:26]: Would normally expect a `?` at the end of the method name here.

File: app/services/user/bulk_destroy.rb [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+    users.all? { |u| guardian.can_delete_user?(u) }
+  end
+
+  def delete(users:, guardian:)
+    users.find_each.with_index do |user, index|
+      position = index + 1
```
Comment (Drenmi) [inline app/services/user/bulk_destroy.rb:32]: You can pass a starting index to `#with_index`:

```suggestion
    users.find_each.with_index(1) do |user, position|
```

File: config/locales/server.en.yml [hunk 19]
```diff
@@ -5534,6 +5535,7 @@ en:
         other: "Automatically revoked, created at more than %{count} days ago"
       revoked: Revoked
       restored: Restored
+    bulk_user_delete: "deleted in a bulk delete operation"
```
Comment (Drenmi) [inline config/locales/server.en.yml:5538]: Is "deleted" intentionally all lowercase here?

**Review by Drenmi — APPROVED** [review]
Excellent PR! Edge cases considered and good test coverage. 👌👨‍🍳

File: config/locales/server.en.yml (hunk 19 above)
Comment (martin-brennan) [reply config/locales/server.en.yml:5538]: Will answer for him here, but yes this is the format of all our staff action logs...it's another remnant of a lot of the Discourse UI being lowercase in the past

File: app/services/user/bulk_destroy.rb (hunk 16 above)
Comment (OsamaSayegh) [reply app/services/user/bulk_destroy.rb:22]: A couple of reason I did this:

1. when I was testing locally, I got a bit confused to see the users being deleted (in the progress logs) in a completely different order to how I selected them
2. in the system specs for the progress logs UI, I need to assert that for each deleted user, we get an entry in the logs. An entry reads like `[X/Y] Successfully deleted @abc` where X is the user position in the list and Y is the total number of users. If the users were retrieved in a random order, the X would be different in each test run.

File: app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js (hunk 14 above)
Comment (OsamaSayegh) [reply app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js:1]: Yeah I agree. I think ideally this controller should be just a wrapper for a component that contains all the logic that currently exists in this controller, but I didn't want to include a fairly large refactoring like that in this PR.

File: app/assets/javascripts/admin/addon/templates/users-list-show.hbs (hunk 15 above)
Comment (OsamaSayegh) [reply app/assets/javascripts/admin/addon/templates/users-list-show.hbs:197]: Yup, the mere presence of the `disabled` attribute makes the input disabled, regardless of its value.

File: app/services/user/bulk_destroy.rb (hunk 17 above)
Comment (OsamaSayegh) [reply app/services/user/bulk_destroy.rb:26]: Yeah if this was a method that we call directly, then a `?` at the end would be appropriate. However, this method name is determined by the `policy` call in the service class definition, and I think `policy :can_delete_users?` doesn't read too well and in all other services we don't use this naming pattern.

File: app/services/user/bulk_destroy.rb (hunk 18 above)
Comment (OsamaSayegh) [reply app/services/user/bulk_destroy.rb]: Nice, TIL! I've made this change.

**Review by martin-brennan — APPROVED** [review]
Okay I think this one can be merged then we can focus on the mobile version :+1:
